[Doc] Add MKL-DNN operator list

TaoLv opened the pull request:
## Description ##
Add MKL-DNN optimized operator list and graph fusion pattern list.

This PR is part of the plan in https://github.com/apache/incubator-mxnet/issues/14399

## Checklist ##
### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

TaoLv commented:
@pengzhao-intel please review~

pengzhao-intel commented on docs/tutorials/mkldnn/operator_list.md:
```diff
@@ -0,0 +1,85 @@
+| FullyConnected + Activation(ReLU) | MXNET_SUBGRAPH_BACKEND  | MXNET_DISABLE_MKLDNN_FUSE_FC_RELU |
+| Convolution (INT8) + re-quantization | MXNET_SUBGRAPH_BACKEND  |   |
+| FullyConnected (INT8) + re-quantization | MXNET_SUBGRAPH_BACKEND  |   |
+
```
The `Enable` column is repeated, I think we can remove this column and only leave the `Disabe` column

TaoLv replied:
Fixed.

pengzhao-intel commented on docs/tutorials/mkldnn/operator_list.md:
```diff
@@ -0,0 +1,85 @@
+| FullyConnected (INT8) + re-quantization | MXNET_SUBGRAPH_BACKEND  |   |
+
+
+To try these features out, you can install MXNet MKL-DNN backend through pip:
```
the install info is already in the below readme page, so I think we can remove in here.

TaoLv replied:
Yes, removed.

pengzhao-intel commented on docs/tutorials/mkldnn/operator_list.md:
```diff
@@ -0,0 +1,85 @@
+
+# MKL-DNN Operator list
+
+MXNet MKL-DNN backend provides optimized implementations for various opertors covering a broad range of applications including image classification, object detection, natural language processing. We also provide the lower precision version for part of these operators on CPU leveraging the DL Boost technology from Intel. On computation graph level, a set of graph fusion pass and quantization pass  ...
```
Make the description as simple as possible?

TaoLv replied:
Changed according to @aaronmarkham 's suggestion.

TaoLv commented:
@xinyu-intel @ciyongch please help to review. Thank you.

xinyu-intel commented on docs/install/index.md:
```diff
@@ -187,7 +187,12 @@ $ pip install mxnet --pre
 </div> <!-- End of master-->
 <hr> <!-- pip footer -->
-MXNet offers MKL pip packages that will be much faster when running on Intel hardware.
+MXNet offers MKL pip packages that will be much faster when running on Intel hardware. Try the following command line to install it and find performance numbers and tuning guide in [performance on Intel CPU](https://mxnet.incubator.apache.org/versions/master/faq/perf.html#intel-cpu).
```
change hyper-link format according to L196

TaoLv replied:
Fixed.

xinyu-intel commented on docs/tutorials/mkldnn/operator_list.md:
```diff
@@ -0,0 +1,83 @@
+
+MXNet MKL-DNN backend provides optimized implementations for various opertors covering a broad range of applications including image classification, object detection, natural language processing. We also provide the lower precision version for part of these operators on CPU leveraging the DL Boost technology from Intel. On computation graph level, a set of graph fusion pass and quantization pass  ...
+
+| Operator           | Function                   | FP32 Training (backward) | FP32 Inference | INT8 Inference |
```
maybe not support markdown table, use text format table instead.

xinyu-intel commented on docs/tutorials/mkldnn/operator_list.md:
````diff
@@ -0,0 +1,83 @@
+export MXNET_DISABLE_MKLDNN_FUSE_CONV_RELU=1
+```
+
+| Fusion pattern                                            | Enable                  | Disable                             |
````
something wrong with table render.

xinyu-intel commented on docs/tutorials/mkldnn/MKLDNN_README.md:
```diff
@@ -19,7 +19,9 @@
 In the following sections, you will find build instructions for MXNet with Intel MKL-DNN on Linux, MacOS and Windows.
-
+
+Please find MKL-DNN optimized operators and other features in [MKL-DNN operator list](http://mxnet.incubator.apache.org/tutorials/mkldnn/operator_list.html)
```
period

TaoLv replied:
Fixed.

anirudhacharya commented:
@mxnet-label-bot add [pr-awaiting-review]

TaoLv commented:
Hi @aaronmarkham May I have your help? CI complains about converting markdown table:
http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/mxnet-validation%2Fwebsite/detail/PR-14891/3/pipeline/85#step-123-log-1335

aaronmarkham commented on docs/install/index.md:
````diff
@@ -124,20 +124,38 @@ Indicate your preferred configuration. Then, follow the customized commands to i
 $ pip install mxnet
 ```
 
+MXNet offers pip packages with MKL-DNN enabled which will be much faster when running on Intel hardware. Try the following command line to install it and find performance numbers and tuning guide in <a href="https://mxnet.incubator.apache.org/versions/master/faq/perf.html#intel-cpu">performance on Intel CPU</a>.
````
```suggestion
MKL-DNN enabled pip packages are optimized for Intel hardware. You can find performance numbers in the <a href="../..//faq/perf.md#intel-cpu">MXNet tuning guide</a>.
```

TaoLv replied:
Hi @aaronmarkham, the link here will be compiled to https://mxnet.incubator.apache.org/versions/faq/perf.md#intel-cpu which is a blank page. Do you think I can use http://mxnet.io/faq/perf.html#intel-cpu instead here?

TaoLv replied:
@pengzhao-intel

aaronmarkham commented on docs/install/index.md:
````diff
@@ -124,20 +124,38 @@ Indicate your preferred configuration. Then, follow the customized commands to i
 $ pip install mxnet==1.3.1
 ```
 
+MXNet offers pip packages with MKL-DNN enabled which will be much faster when running on Intel hardware. Try the following command line to install it and find performance numbers and tuning guide in <a href="https://mxnet.incubator.apache.org/versions/master/faq/perf.html#intel-cpu">performance on Intel CPU</a>.
````
```suggestion
MKL-DNN enabled pip packages are optimized for Intel hardware. You can find performance numbers in the <a href="../../faq/perf.md#intel-cpu">MXNet tuning guide</a>.
```

aaronmarkham commented on docs/install/index.md:
````diff
@@ -124,20 +124,38 @@ Indicate your preferred configuration. Then, follow the customized commands to i
 $ pip install mxnet==1.2.1
 ```
 
+MXNet offers pip packages with MKL-DNN enabled which will be much faster when running on Intel hardware. Try the following command line to install it and find performance numbers and tuning guide in <a href="https://mxnet.incubator.apache.org/versions/master/faq/perf.html#intel-cpu">performance on Intel CPU</a>.
````
```suggestion
MKL-DNN enabled pip packages are optimized for Intel hardware. You can find performance numbers in the <a href="../../faq/perf.md#intel-cpu">MXNet tuning guide</a>.
```

aaronmarkham commented on docs/install/index.md:
````diff
@@ -185,9 +203,15 @@ $ pip install mxnet==0.11.0
 $ pip install mxnet --pre
 ```
 
+MXNet offers pip packages with MKL-DNN enabled which will be much faster when running on Intel hardware. Try the following command line to install it and find performance numbers and tuning guide in <a href="https://mxnet.incubator.apache.org/versions/master/faq/perf.html#intel-cpu">performance on Intel CPU</a>.
````
```suggestion
MKL-DNN enabled pip packages are optimized for Intel hardware. You can find performance numbers in the <a href="../../faq/perf.md#intel-cpu">MXNet tuning guide</a>.
```

aaronmarkham commented on docs/tutorials/mkldnn/MKLDNN_README.md:
```diff
@@ -19,7 +19,9 @@
 In the following sections, you will find build instructions for MXNet with Intel MKL-DNN on Linux, MacOS and Windows.
-
+
+Please find MKL-DNN optimized operators and other features in [MKL-DNN operator list](http://mxnet.incubator.apache.org/tutorials/mkldnn/operator_list.html).
```
```suggestion
Please find MKL-DNN optimized operators and other features in [MKL-DNN operator list](../mkldnn/operator_list.md).
```

TaoLv commented:
Thank you for the great suggestions, @aaronmarkham .

ciyongch commented on docs/tutorials/mkldnn/operator_list.md:
````diff
@@ -0,0 +1,79 @@
+For example, you can enable all fusion passes by:
+
+```
+export MXNET_SUBGRAPH_BACKEND=MKLDNN
````
`all fusion passes` might not be accurate enough here.
`MKLDNN` backend is only applicable to the FP32 mode fusion, while `MKLDNN_POST_QUANTIZE` is applicable to the INT8 mode fusion (quantized_op +requantize/dequantize).

TaoLv replied:
Nice to know! How many subgraph backend options do we have now? What about INT8 convolution + relu fusion?

ciyongch replied:
Currently, there're only two options as mentioned above.
There's no fusion pass for `INT8 conv +relu`, the current flow is doing the fusion of `FP32 conv +relu` firstly, and then quantizing `FP32 sg_conv` to `INT8 sg_conv`.

aaronmarkham replied:
> Thank you for the information. I think we need to put them into this page: https://mxnet.incubator.apache.org/versions/master/faq/env_var.html

Let's maintain them in one spot and then reference where appropriate. Either way some mention is needed in the env_var page.

ciyongch commented on docs/tutorials/mkldnn/operator_list.md:
````diff
@@ -0,0 +1,79 @@
+
+```
+export MXNET_DISABLE_MKLDNN_FUSE_CONV_RELU=1
+```
````
There's also other options for end-use to choose, such as:
`MXNET_DISABLE_MKLDNN_CONV_OPT` to disable all MKLDNN convolution optimization pass.
`MXNET_DISABLE_MKLDNN_FC_OPT` to disable all MKLDNN FullyConnected optimization pass.

ciyongch commented on docs/tutorials/mkldnn/operator_list.md:
```diff
@@ -0,0 +1,79 @@
+| FullyConnected + Activation(ReLU)                         | MXNET_DISABLE_MKLDNN_FUSE_FC_RELU   |
+| Convolution (INT8) + re-quantization                      |                                     |
+| FullyConnected (INT8) + re-quantization                   |                                     |
+
```
miss a pattern here: `FullyConnected (INT8) + re-quantization + de-quantization`

TaoLv commented:
@aaronmarkham @ciyongch @pengzhao-intel please take another look - I added some description about how the subgraph backend variable is used and added a reference link in the env var page.

aaronmarkham commented on docs/faq/env_var.md:
````diff
@@ -280,6 +280,11 @@ When USE_PROFILER is enabled in Makefile or CMake, the following environments ca
+* MXNET_SUBGRAPH_BACKEND
+  - Values: String ```(default="")```
+  - This variable controls the subgraph partitioning in MXNet.
+  - This variable is used to perform MKL-DNN FP32 operator fusion and quantization. Please refer to [MKL-DNN operator list](../tutorials/mkldnn/operator_list.md) for how this variable is used and the list of fusion pass.
````
```suggestion
  - This variable is used to perform MKL-DNN FP32 operator fusion and quantization. Please refer to the [MKL-DNN operator list](../tutorials/mkldnn/operator_list.md) for how this variable is used and the list of fusion passes.
```

aaronmarkham commented on docs/install/index.md:
````diff
@@ -124,20 +124,38 @@ Indicate your preferred configuration. Then, follow the customized commands to i
 $ pip install mxnet
 ```
 
+MKL-DNN enabled pip packages are optimized for Intel hardware. You can find performance numbers in the <a href="../..//faq/perf.md#intel-cpu">MXNet tuning guide</a>.
````
extra slash for a reason? that might have been my mistake...
```suggestion
MKL-DNN enabled pip packages are optimized for Intel hardware. You can find performance numbers in the <a href="../../faq/perf.md#intel-cpu">MXNet tuning guide</a>.
```

aaronmarkham commented on docs/tutorials/mkldnn/MKLDNN_README.md:
```diff
@@ -1,25 +1,27 @@
 A better training and inference performance is expected to be achieved on Intel-Architecture CPUs with MXNet built with [Intel MKL-DNN](https://github.com/intel/mkl-dnn) on multiple operating system, including Linux, Windows and MacOS.
 In the following sections, you will find build instructions for MXNet with Intel MKL-DNN on Linux, MacOS and Windows.
 
+Please find MKL-DNN optimized operators and other features in [MKL-DNN operator list](../mkldnn/operator_list.md).
```
```suggestion
Please find MKL-DNN optimized operators and other features in the [MKL-DNN operator list](../mkldnn/operator_list.md).
```

TaoLv commented:
Thank you for the fix @aaronmarkham . I'm not familiar with the compilation from md files to html files, but I not notice there is already an index.md in the folder (and other folders under tutorials). Are you suggesting to overwrite the existing index.md?

aaronmarkham commented:
> Thank you for the fix @aaronmarkham . I'm not familiar with the compilation from md files to html files, but I not notice there is already an index.md in the folder (and other folders under tutorials). Are you suggesting to overwrite the existing index.md?

Since others are working on that file right now in another PR, let's leave it alone for now.

pengzhao-intel commented:
@TaoLv I merged two MKLDNN related PRs just now and please check if the doc needs the related change.
I think we can merge this PR when all is finalized.

TaoLv commented:
> @TaoLv I merged two MKLDNN related PRs just now and please check if the doc needs the related change.
> I think we can merge this PR when all is finalized.

Sure. Will change accordingly~

TaoLv commented:
@pengzhao-intel  Change the environmental variable name from `MKLDNN_POST_QUANTIZE` to `MKLDNN_QUANTIZE` accordingly. And also remove the `ReLU` constraint since we support more activation fusion now. Please review again~

pengzhao-intel commented:
It's great @TaoLv, please retrigger the CI and I will merge soon.

pengzhao-intel commented:
update `env_var.md` again :(
